docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,552 @@
+    ```go
+    // Current Code
+    // sdk.go.tpl    
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 false }}
+        ...
+    }
+    ```
+
+    ```go
+    // Proposed Code
+    // sdk.go.tpl
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        // NOTE: The performSpecUpdate boolean param will be removed
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 }}
+        ...
+    }
+    ```
+
+2. **Provide a new "LateInitializationConfig" struct inside generator config for AWS service teams to provide which spec fields will be late initialized.** These fields will be the ones which are late initialized by the server but not present in the Create and Update response.
+
+    These fields will be populated by making the readOne call. AWS service teams can provide the configuration to either late initialize these fields immidiately after Create/Update call or provide a delay after which these fields will be late initialized.
+    LateInitializationConfig will be added as part of FieldConfig. 
+
+    ```go
+        // LateInitializeConfig contains instructions for how to handle the
+        // retrieval and setting of server-side defaulted fields.
+        type LateInitializationConfig struct {
+            // DefaultImmediately indicates that if the user has not
+            // specified a value for the field, then directly after a
+            // successful call to Create/Update, the reconciler will call
+            // ReadOne to get the value of the field as defaulted by the
+            // AWS service, set the value of this field on the resource and
+            // patch the resource back to the Kubernetes API.
+            DefaultImmidiately bool `json:"default_immidiately"`
+            // DelaySeconds provides the number of seconds to wait to set
+            // late initialized fields after successful Create/Update call
+            // DelaySeconds is only used if DefaultImmidiately is false
+            // Default value for DelaySeconds is 5 seconds if DefaultImmidiately
+            // is false and DelaySeconds is absent.
+            DelaySeconds *int `json:"delay_seconds,omitempty"`
+        }
+
+        type FieldConfig struct {
+            LateInitialize *LateInitializeConfig `json:"late_initialize,omitempty"`
+        }
+    ```
+
+3. **Provide 3 new hooks to AWS service teams to customize the code generation in step 4**
+    
+    These 3 hooks will be:
+    
+    a) **set_late_initialized_defaults_override**: This hook will provide the **complete implementation** for setting default values of optional spec fields which are not provided by the k8s user and defaulted by AWS service.
```

Comment:
we don't use hooks for code that replaces entire methods (custom overrides). Hooks represent places within the regular code flow or code file that are executed if present. Instead, we use fields on a configuration struct to indicate that custom code entirely replaces a method.

For example, the [`pkg/generate/config.OperationConfig:CustomImplementation`](https://github.com/aws-controllers-k8s/code-generator/blob/c13e5fe097c1b487ac54d884761178e3cdeac3a7/pkg/model/crd.go#L479-L495) field is [used within the templates](https://github.com/aws-controllers-k8s/code-generator/blob/c13e5fe097c1b487ac54d884761178e3cdeac3a7/templates/pkg/resource/sdk.go.tpl#L57-L62) to entirely replace a particular implementation of a method.